### PR TITLE
[tests-only] retry if auto-sync is enabled but client.synchronize is false

### LINKED ISSUE #8859: Auto-accepting is slow in CI
https://drone.owncloud.com/owncloud/ocis/33724/32/6
https://drone.owncloud.com/owncloud/ocis/33662/27/5
https://drone.owncloud.com/owncloud/ocis/33734/32/6

Steps: 
- user1 shares file 
- user2 send `GET https://localhost:9200/graph/v1beta1/me/drive/sharedWithMe`

Actual: share should be accepted. `@client.synchronize: true`

I have not been able to reproduce this locally. I don't see any problems on the testing side. adding a waiting to the test may solve the problem but seems like a poor solution to the problem

### LINKED ISSUE #8864: [QA] Flaky API test
## Describe the bug
API tests are failing because `AfterScenario` after test `tests/acceptance/features/apiSharingNg/sharedByMe.feature:1487` is failing which is causing the next test i.e. `tests/acceptance/features/apiSharingNg/sharedWithMe.feature:15` to fail

### Builds
https://drone.owncloud.com/owncloud/ocis/33717/32/5
https://drone.owncloud.com/owncloud/ocis/33719/32/5
https://drone.owncloud.com/owncloud/ocis/33720/32/5

```feature
# previous scenario fail

  @env-config
  Scenario: user lists shared resources for deleted group                                           # /drone/src/tests/acceptance/features/apiSharingNg/sharedByMe.feature:1487
    Given the config "GRAPH_SPACES_GROUPS_CACHE_TTL" has been set to "1"                            # OcisConfigContext::theConfigHasBeenSetTo()
    And group "grp1" has been created                                                               # FeatureContext::groupHasBeenCreated()
    And the administrator has assigned the role "Admin" to user "Alice" using the Graph API         # GraphContext::theAdministratorHasGivenTheRoleUsingTheGraphApi()
    And user "Alice" has uploaded file with content "hello world" to "textfile.txt"                 # FeatureContext::userHasUploadedAFileWithContentTo()
    And user "Alice" has sent the following share invitation:                                       # SharingNgContext::userHasSentTheFollowingShareInvitation()
      | resource        | textfile.txt |
      | space           | Personal     |
      | sharee          | grp1         |
      | shareType       | group        |
      | permissionsRole | Viewer       |
    And group "grp1" has been deleted                                                               # FeatureContext::groupHasBeenDeleted()
    When user "Alice" lists the shares shared by her after clearing group cache using the Graph API # GraphContext::userListsTheResourcesSharedByAUserAfterClearingUserOrGroupSpaceUsingGraphApi()
    Then the HTTP status code should be "200"                                                       # FeatureContext::thenTheHTTPStatusCodeShouldBe()
    And the JSON data of the response should match                                                  # FeatureContext::theJsonDataOfTheResponseShouldMatch()
      """
      {
        "type": "object",
        "required": [
          "value"
        ],
        "properties": {
          "value": {
            "type": "array",
            "minItems":0,
            "maxItems":0
          }
        }
      }
      """
  │
  ╳  HTTP status code 404 is not the expected value 204
  ╳  Failed asserting that 404 matches expected 204.
  │
  └─ @AfterScenario # FeatureContext::afterScenario()

# apiSharingNg/sharedWithMe.feature:15
Scenario: user lists the file shared with them                                        # /drone/src/tests/acceptance/features/apiSharingNg/sharedWithMe.feature:15
    Given user "Alice" has uploaded file with content "hello world" to "/textfile0.txt" # FeatureContext::userHasUploadedAFileWithContentTo()
    And user "Alice" has sent the following share invitation:                           # SharingNgContext::userHasSentTheFollowingShareInvitation()
      | resource        | textfile0.txt |
      | space           | Personal      |
      | sharee          | Brian         |
      | shareType       | user          |
      | permissionsRole | Viewer        |
    When user "Brian" lists the shares shared with him using the Graph API              # GraphContext::userListsTheResourcesSharedWithThemUsingGraphApi()
    Then the HTTP status code should be "200"                                           # FeatureContext::thenTheHTTPStatusCodeShouldBe()
    And the JSON data of the response should match
     Enum failed, enum: [true], data: false at #->properties:value->items[0]:0->properties:@client.synchronize (Swaggest\JsonSchema\Exception\EnumException)

# response
{
  "value": [
    {
      "@UI.Hidden": false,
      "@client.synchronize": false
      ...
    }
  ]
}
```

### PR DESCRIPTION
**Pull request by saw-jan**
## Description
Implemented retry when listing the shares if auto-sync is enabled but client.synchronize is false. If the auto-sync is disabled then the retry will be skipped.

Scenarios can pass with retry:
```feature
    When user "Brian" lists the shares shared with him using the Graph API
      │ auto-sync share for user 'Brian' is enabled
      │ but share 'textfile0.txt' was not auto-synced, retrying (1)...
      │ 
    Then the HTTP status code should be "200"
```
## Related Issue
- Revert https://github.com/owncloud/ocis/commit/4e66cc730e633cd3ff3dac787e9bd5f92c0b7224
- Fixes #8864, Fixes #8859

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Comment by saw-jan** [comment]
Want to check the edge cases

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 1]
```diff
@@ -2505,14 +2505,38 @@ public function userListsTheResourcesSharedWithThemUsingGraphApi(string $user, s
...
+			$response = GraphHelper::getSharesSharedWithMe(
 				$this->featureContext->getBaseUrl(),
 				$this->featureContext->getStepLineRef(),
 				$credentials['username'],
 				$credentials['password']
-			)
-		);
+			);
+
+			$jsonObj = $this->featureContext->getJsonDecodedResponseBodyContent($response);
```
Comment (amrita-shrestha) [inline tests/acceptance/features/bootstrap/GraphContext.php:2522]:
```suggestion
			$jsonObj = $this->featureContext->getJsonDecodedResponseBodyContent($response);
```
$jsonObj name should be replace with meaningful name

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 2]
```diff
@@ -2505,14 +2505,38 @@ public function userListsTheResourcesSharedWithThemUsingGraphApi(string $user, s
...
+				if ($tryAgain) {
+					$retried += 1;
+					echo "auto-sync share for user '$user' is enabled\n";
+					echo "but share '$share->name' was not auto-synced, retrying ($retried)...\n";
+					// wait 500ms and try again
+					\usleep(500 * 1000);
+					break;
+				}
+			}
+		} while ($tryAgain);
```
Comment (amrita-shrestha) [inline tests/acceptance/features/bootstrap/GraphContext.php:2537]:
```suggestion
		} while ($tryAgain);
```
isn't it dangerous to use. I think we should limit tryAgain ?

Comment (saw-jan) [reply tests/acceptance/features/bootstrap/GraphContext.php:2537]: this line should prevent the infinite loop
```php
$tryAgain = !$share->{'@client.synchronize'} && $autoSync && $retried < HttpRequestHelper::numRetriesOnHttpTooEarly();
```

**Comment by saw-jan** [comment]
Previously flaky scenarios are consistent with this explicit wait https://github.com/owncloud/ocis/commit/4e66cc730e633cd3ff3dac787e9bd5f92c0b7224. Though it is not so good to have hard coded wait, the fix in this PR (retry step) is also not that much promising because it can lead to confusion in the future where there might be scenarios that manipulate share-sync config in different ways.

So, let's keep the wait fix and I will close this PR for now.
If anyone has any though then please let me know.

CC @ScharfViktor

**Comment by ScharfViktor** [comment]
to avoid flaky we can check that `sychronize` is true using step with timeout `lists the shares shared with him after clearing user cache using the Graph API`


for another scenarios we can check it like `true,false`: 

```
"@client.synchronize": {
                "type": "boolean",
                "enum": [true,false]
              },
```

**Comment by saw-jan** [comment]
Let's keep the current fix. We can come back to this if we get other flakiness

File: tests/acceptance/features/bootstrap/FeatureContext.php [hunk 3]
```diff
@@ -167,6 +168,46 @@ class FeatureContext extends BehatVariablesContext {
...
+	 * @return void
+	 */
+	public function rememberUserAutoSyncSetting(string $user, bool $value): void {
+		$this->autoSyncSettings[$user] = $value;
+	}
+
+	/**
+	 * this is set true for db conversion tests
+	 */
+	private bool $dbConversion = false;
```
Comment (ScharfViktor) [inline tests/acceptance/features/bootstrap/FeatureContext.php:209]: what do we need this for?

File: tests/TestHelpers/SettingsHelper.php [hunk 4]
```diff
@@ -0,0 +1,281 @@
+<?php declare(strict_types=1);
+/**
+ * ownCloud
+ *
+ * @author Sajan Gurung <sajan@jankaritech.com>
+ * @copyright Copyright (c) 2023 Sajan Gurung sajan@jankaritech.com
```
Comment (ScharfViktor) [inline tests/TestHelpers/SettingsHelper.php:6]: 2024

**Review by ScharfViktor — APPROVED** [review]
LGTM

File: tests/acceptance/features/bootstrap/FeatureContext.php (hunk 3 above)
Comment (saw-jan) [reply tests/acceptance/features/bootstrap/FeatureContext.php:209]: that should not be there. :+1:
